File: php/app/service/AuthService.php
<?php
declare(strict_types=1);

namespace app\service;

use RuntimeException;
use think\facade\Db;
use think\Request;

class AuthService
{
    public function login(string $username, string $password, Request $request): array
    {
        $username = trim($username);
        if ($username === '' || $password === '') {
            throw new RuntimeException('请输入账号和密码');
        }

        $user = Db::name('admin_users')->where('username', $username)->find();
        if (!$user || (int) $user['status'] !== 1 || !password_verify($password, (string) $user['password_hash'])) {
            throw new RuntimeException('账号或密码不正确');
        }

        $token = bin2hex(random_bytes(32));
        $expiresAt = date('Y-m-d H:i:s', time() + 7 * 86400);
        $now = date('Y-m-d H:i:s');

        Db::name('admin_users')->where('id', $user['id'])->update([
            'auth_token_hash' => hash('sha256', $token),
            'token_expires_at' => $expiresAt,
            'last_login_at' => $now,
            'last_login_ip' => $request->ip(),
            'updated_at' => $now,
        ]);

        return [
            'token' => $token,
            'expiresAt' => $expiresAt,
            'user' => $this->publicUser($user),
        ];
    }

    public function authenticate(Request $request): ?array
    {
        $token = $this->bearerToken($request);
        if ($token === null) {
            return null;
        }

        $user = Db::name('admin_users')
            ->where('auth_token_hash', hash('sha256', $token))
            ->where('status', 1)
            ->find();

        if (!$user) {
            return null;
        }

        $expiresAt = strtotime((string) ($user['token_expires_at'] ?? ''));
        if (!$expiresAt || $expiresAt < time()) {
            return null;
        }

        return $user;
    }

    public function logout(Request $request): void
    {
        $user = $this->authenticate($request);
        if (!$user) {
            return;
        }

        Db::name('admin_users')->where('id', $user['id'])->update([
            'auth_token_hash' => null,
            'token_expires_at' => null,
            'updated_at' => date('Y-m-d H:i:s'),
        ]);
    }

    public function publicUser(array $user): array
    {
        return [
            'id' => (int) $user['id'],
            'username' => (string) $user['username'],
            'displayName' => (string) $user['display_name'],
            'role' => (string) $user['role'],
        ];
    }

    private function bearerToken(Request $request): ?string
    {
        $header = trim((string) $request->header('authorization', ''));
        if ($header === '' || !preg_match('/^Bearer\s+(.+)$/i', $header, $matches)) {
            return null;
        }

        $token = trim($matches[1]);
        return $token === '' ? null : $token;
    }
}

File: java/src/main/java/com/molizhishu/demo/service/AuthService.java
package com.molizhishu.demo.service;

import com.molizhishu.demo.mapper.AuthMapper;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

@Service
public class AuthService {

	private final AuthMapper            authMapper;

	private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

	public AuthService(AuthMapper authMapper) {
		this.authMapper = authMapper;
	}

	public Map<String, Object> login(String username, String password, HttpServletRequest request) {
		Map<String, Object> user = authMapper.findByUsername(username == null ? "" : username.trim());
		if (user == null || password == null || password.isBlank() || !passwordEncoder.matches(password, String.valueOf(user.get("password_hash")))) {
			throw new IllegalArgumentException("账号或密码不正确");
		}

		String        token     = UUID.randomUUID().toString().replace("-", "") + UUID.randomUUID().toString().replace("-", "");
		LocalDateTime expiresAt = LocalDateTime.now().plusDays(7);
		authMapper.updateToken(user.get("id"), sha256(token), expiresAt, request.getRemoteAddr());

		return Map.of("token", token, "expiresAt", expiresAt.toString().replace("T", " "), "user", publicUser(user));
	}

	public Map<String, Object> authenticate(String token) {
		if (token == null || token.isBlank()) {
			return null;
		}

		Map<String, Object> user = authMapper.findByTokenHash(sha256(token.trim()));
		if (user == null) {
			return null;
		}

		Object expiresAt = user.get("token_expires_at");
		if (expiresAt instanceof LocalDateTime time && time.isBefore(LocalDateTime.now())) {
			return null;
		}

		return user;
	}

	public void logout(String token) {
		Map<String, Object> user = authenticate(token);

		if (user != null) {
			authMapper.clearToken(user.get("id"));
		}
	}

	public Map<String, Object> publicUser(Map<String, Object> user) {
		Map<String, Object> result = new LinkedHashMap<>();
		result.put("id", user.get("id"));
		result.put("username", user.get("username"));
		result.put("displayName", user.get("display_name"));
		result.put("role", user.get("role"));

		return result;
	}

	public String bearerToken(HttpServletRequest request) {
		String header = request.getHeader("Authorization");
		if (header == null || !header.regionMatches(true, 0, "Bearer ", 0, 7)) {
			return "";
		}
		return header.substring(7).trim();
	}

	private String sha256(String value) {
		try {
			return HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(value.getBytes(StandardCharsets.UTF_8)));
		} catch (NoSuchAlgorithmException e) {
			throw new IllegalStateException(e);
		}
	}

}
